api/src/modules/indicator-records/indicator-records.service.ts

```diff
@@ -45,12 +53,67 @@ export class IndicatorRecordsService extends AppBaseService<
       keyForAttribute: 'camelCase',
     };
   }
+
   async getIndicatorRecordById(id: number): Promise<IndicatorRecord> {
-    const found = await this.indicatorRecordRepository.findOne(id);
+    const found:
+      | IndicatorRecord
+      | undefined = await this.indicatorRecordRepository.findOne(id);
     if (!found) {
       throw new NotFoundException(`Indicator Record with ID "${id}" not found`);
     }
 
     return found;
   }
+
+  async calculateImpactValue(
+    sourcingRecord: SourcingRecord,
+  ): Promise<IndicatorRecord[]> {
+    this.logger.log(
+      `Calculating impact value for sourcing record ${sourcingRecord.id}`,
+    );
+    const indicators: Indicator[] = await this.indicatorService.findAllUnpaginated();
+
+    if (!sourcingRecord.sourcingLocation.geoRegion) {
+      throw new Error(
+        'Cannot calculate impact for sourcing record - missing geoRegion (through sourcingLocation)',
+      );
+    }
+
+    if (!sourcingRecord.sourcingLocation.material) {
+      throw new Error(
+        'Cannot calculate impact for sourcing record - missing material (through sourcingLocation)',
+      );
+    }
+    const productionValue: number = await this.h3DataService.getProductionForGeoRegion(
+      sourcingRecord.sourcingLocation.geoRegion,
+      sourcingRecord.sourcingLocation.material,
+    );
+
+    const indicatorRecords: IndicatorRecord[] = await Promise.all(
+      indicators.map(
+        async (indicator: Indicator): Promise<IndicatorRecord> => {
+          const impactValue: number = await this.h3DataService.getImpactForGeoRegion(
+            sourcingRecord.sourcingLocation.geoRegion,
+            indicator,
+          );
+
+          return IndicatorRecord.merge(new IndicatorRecord(), {
+            value: (impactValue * sourcingRecord.tonnage) / productionValue,
+            indicatorId: indicator.id,
+            status: INDICATOR_RECORD_STATUS.SUCCESS,
+            sourcingRecordId: sourcingRecord.id,
+          });
+        },
+      ),
+    );
+
+    this.logger.log(
+      `Impact value for sourcing record ${sourcingRecord.id} saved.`,
+    );
+    return this.indicatorRecordRepository.save(indicatorRecords);
```

Comment:
Just FYI for future optimizations; typeorm `save` is inefficient, as it does two queries first checking if the record exists then either insert or update. If you know which records to update, better to do a bulk delete, then bulk insert. Similarly, some queries could probably be batched in chunks.